src/storage/src/hummock/event_handler/uploader.rs

```diff
@@ -523,48 +503,306 @@ impl SealedData {
-            }
             self.spilled_data.add_task(task);
-            size
+        }
+    }
+}
+
+struct LocalInstanceEpochData {
+    epoch: HummockEpoch,
+    // newer data comes first.
+    imms: VecDeque<UploaderImm>,
+    has_spilled: bool,
+}
+
+impl LocalInstanceEpochData {
+    fn new(epoch: HummockEpoch) -> Self {
+        Self {
+            epoch,
+            imms: VecDeque::new(),
+            has_spilled: false,
+        }
+    }
+
+    fn epoch(&self) -> HummockEpoch {
+        self.epoch
+    }
+
+    fn add_imm(&mut self, imm: UploaderImm) {
+        assert_eq!(imm.max_epoch(), imm.min_epoch());
+        assert_eq!(self.epoch, imm.min_epoch());
+        if let Some(prev_imm) = self.imms.front() {
+            assert_gt!(imm.batch_id(), prev_imm.batch_id());
+        }
+        self.imms.push_front(imm);
+    }
+
+    fn is_empty(&self) -> bool {
+        self.imms.is_empty()
+    }
+}
+
+struct LocalInstanceUnsyncData {
+    table_id: TableId,
+    instance_id: LocalInstanceId,
+    // None means that the current instance should have stopped advancing
+    current_epoch_data: Option<LocalInstanceEpochData>,
+    // newer data comes first.
+    sealed_data: VecDeque<LocalInstanceEpochData>,
+    // newer data comes first
+    flushing_imms: VecDeque<SharedBufferBatchId>,
+}
+
+impl LocalInstanceUnsyncData {
+    fn new(table_id: TableId, instance_id: LocalInstanceId, init_epoch: HummockEpoch) -> Self {
+        Self {
+            table_id,
+            instance_id,
+            current_epoch_data: Some(LocalInstanceEpochData::new(init_epoch)),
+            sealed_data: VecDeque::new(),
+            flushing_imms: Default::default(),
+        }
+    }
+
+    fn add_imm(&mut self, imm: UploaderImm) {
+        assert_eq!(self.table_id, imm.table_id);
+        self.current_epoch_data
+            .as_mut()
+            .expect("should be Some when adding new imm")
+            .add_imm(imm);
+    }
+
+    fn local_seal_epoch(&mut self, next_epoch: HummockEpoch) -> HummockEpoch {
+        let data = self
+            .current_epoch_data
+            .as_mut()
+            .expect("should be Some when seal new epoch");
+        let current_epoch = data.epoch;
+        debug!(
+            instance_id = self.instance_id,
+            next_epoch, current_epoch, "local seal epoch"
+        );
+        assert_gt!(next_epoch, current_epoch);
+        let epoch_data = replace(data, LocalInstanceEpochData::new(next_epoch));
+        if !epoch_data.is_empty() {
+            self.sealed_data.push_front(epoch_data);
+        }
+        current_epoch
+    }
+
+    // imm_ids from old to new, which means in ascending order
+    fn ack_flushed(&mut self, imm_ids: impl Iterator<Item = SharedBufferBatchId>) {
+        for imm_id in imm_ids {
+            assert_eq!(self.flushing_imms.pop_back().expect("should exist"), imm_id);
+        }
+    }
+
+    fn spill(&mut self, epoch: HummockEpoch) -> Vec<UploaderImm> {
```

Comment:
minor: shall we return `Option<Vec<UploaderImm>>` for `fn spill` and `fn sync` so that we can save `VecDeque` initilization (which requires memory allocation) when there is nothing to spill/sync?

Reply:
`VecDeque::new` is a const fn, which does not involve memory allocation. ditto for `Vec::new`.